cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/store/DefaultMetadataStore.java

```diff
@@ -552,16 +555,46 @@ private MetadataDataset newMetadataDataset(MetadataScope scope) {
    * @param framework Dataset framework to add types and datasets to
    */
   public static void setupDatasets(DatasetFramework framework) throws IOException, DatasetManagementException {
-    framework.addInstance(MetadataDataset.class.getName(), BUSINESS_METADATA_INSTANCE_ID, DatasetProperties.EMPTY);
+    // In 3.2 we used MetadataDataset.KEYVALUE_COLUMN (kv) and MetadataDataset.CASE_INSENSITIVE_VALUE_COLUMN (civ) and
+    // in 3.3 we index new column MetadataDataset.CASE_INSENSITIVE_VALUE_COLUMN (i). During upgrade we want to
+    // have instance of BUSINESS_METADATA_INSTANCE_ID with columnToIndex with all these three column so that
+    // when we do a delete while upgrading MetadataDataset#upgrade the delete operation also delete the records
+    // for the old index column from index table.
+    DatasetProperties dsProperties = getOldColumnToIndex();
+    // the above is only needed for Business Metadata as in 3.2 we didn't have System Metadata table.
+    //TODO: (UPG-3.3): Remove this after 3.3
+    framework.addInstance(MetadataDataset.class.getName(), BUSINESS_METADATA_INSTANCE_ID, dsProperties);
     framework.addInstance(MetadataDataset.class.getName(), SYSTEM_METADATA_INSTANCE_ID, DatasetProperties.EMPTY);
   }
 
   public void upgrade() {
-    execute(new TransactionExecutor.Procedure<MetadataDataset>() {
+    DatasetProperties dsProperties = getOldColumnToIndex();
+    MetadataDataset businessMetadataDataset;
+    try {
+      businessMetadataDataset = DatasetsUtil.getOrCreateDataset(
+        dsFramework, BUSINESS_METADATA_INSTANCE_ID, MetadataDataset.class.getName(),
+        dsProperties, DatasetDefinition.NO_ARGUMENTS, null);
+    } catch (DatasetManagementException | IOException e) {
+      throw Throwables.propagate(e);
+    }
+    Preconditions.checkNotNull(businessMetadataDataset, "Failed to get Business Metadata Dataset.");
+    TransactionExecutor txExecutor = Transactions.createTransactionExecutor(txExecutorFactory, businessMetadataDataset);
+      txExecutor.executeUnchecked(new TransactionExecutor.Procedure<MetadataDataset>() {
       @Override
       public void apply(MetadataDataset input) throws Exception {
         input.upgrade();
       }
-    }, MetadataScope.USER); // in 3.2 we only had user metadata so upgrade only that
+    }, businessMetadataDataset);
+  }
+
+  /**
+   * @return {@link DatasetProperties} where {@link IndexedTableDefinition#INDEX_COLUMNS_CONF_KEY} is set to
+   * index columns from 3.2 and also 3.3
+   */
+  private static DatasetProperties getOldColumnToIndex() {
```

Comment:
if this method returns ``DatasetProperties`` should we call it ``getOldDatasetProperties``?